96.md

```diff
@@ -0,0 +1,221 @@
+
+File storage servers wishing to be accessible by nostr users should opt-in by making available an https route at `/.well-known/nostr/nip96.json` with `api_url`:
+
+```js
+{
+  // Required
+  // File upload and deletion are served from this url
+  // Also downloads if "download_url" field is absent or empty string
+  "api_url": "https://your-file-server.example/custom-api-path",
+  // Optional
+  // If absent, downloads are served from the api_url
+  "download_url": "https://a-cdn.example/a-path",
+  // Optional
+  // Use this if you are a nostr relay using your /.well-known/nostr/nip96.json
+  // just to advertise someone else's http file storage server's /.well-known/nostr/nip96.json
+  // In this case, "api_url" field must be an empty string
+  "recommended_url": "https://your-file-server.example",
+  // Optional
+  "supported_nips": [],
+  // Optional
+  "tos_url": "https://your-file-server.example/terms-of-service",
+  // Optional
+  "plans": {
+    // "free" is the only standardized plan key and
+    // clients may use its presence to learn if server offers free storage
+    "free": {
+      "name": "Free Tier",
+      "url": "https://...", // payment etc
+      "max_byte_size": 10485760,
+      // Range in days / 0 for no expiration
+      // [7, 0] means it may vary from 7 days to unlimited persistence,
+      // [0, 0] means it has no expiration
+      // early expiration may be due to low traffic or any other factor
+      "file_expiration": [14, 90],
+      "image_transformations": ["resizing"]
+    }
+  }
+}
+```
+
+Note that if using a CDN, the `server` MAY issue a 302 redirect from
+GET requests to the `api_url` or `download_url` route to the CDN url instead
+of assigning the CDN directly as the `download_url` route.
+This is because some NIP-96 integrations
+MAY require NIP-98 authentication for downloads,
+which the CDN won't be able to support.
+
+### Relay Hints
+
+A nostr relay MAY advertise someone else's HTTP file storage server by
+adding a `/.well-known/nostr/nip96.json` with "recommended_url" field
+pointing to the url where the server hosts its own
+`/.well-known/nostr/nip96.json`. In this case, the "api_url" must
+be an empty string.
+
+If the nostr relay is also an HTTP file storage server,
+it must use the "api_url" field instead.
+
+### List of Supporting File Storage Servers
+
+| Name | Domain |
+| ---- | ------ |
+| -    | -      |
+
+## Upload
+
+A file can be uploaded one at a time to `https://your-file-server.example/custom-api-path` (route from `https://your-file-server.example/.well-known/nostr/nip96.json` "api_url" field) as `multipart/form-data` content type using `POST` method with the file object set to the `file` form data field.
+
+`Clients` must add a [NIP-98](98.md) `Authorization` header with the encoded `payload` tag set to the base64-encoded 256-bit SHA-256 hash of the file (not the hash of the whole request body).
+If using an html form, use an `Authorization` form data field instead.
```

Comment:
If using an HTML `<form>` to send file without javascript, the client won't be able to set request headers. That's why the NIP-98 `Authorization` content will be a regular field instead of a header.